sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -609,6 +610,19 @@ object CollapseWindow extends Rule[LogicalPlan] {
 }
 
 /**
+ * Transpose Adjacent Window Expressions.
+ * - If the partition spec of the parent Window expression is a subset of the partition spec
+ *   of the child window expression, transpose them.
+ */
+object TransposeWindow extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if ps1.length < ps2.length && ps2.containsSlice(ps1) =>
```

Comment:
We might be able to get a little more milage out of the rule by using `semanticEquals` for comparing the partition expressions, e.g.:
```scala
def sliceSemanticEquals(ps1: Seq[Expression], ps2: Seq[Expression]): Boolean = ps1.zip(ps2).forall {
  case (l, r) => l.semanticEquals(r)
}
...
sliceSemanticEquals(ps1, ps2)
``` 

You could even get more leverage if you do not consider the order of the partition spec.